File: features/transaction/checkout/src/main/java/com/tokopedia/checkout/view/feature/emptycart/viewmodel/RecentViewViewModel.java
package com.tokopedia.checkout.view.feature.emptycart.viewmodel;

import android.os.Parcel;
import android.os.Parcelable;

import com.tokopedia.checkout.domain.datamodel.recentview.RecentView;

/**
 * Created by Irfan Khoirul on 21/09/18.
 */

public class RecentViewViewModel implements Parcelable {

    private RecentView recentView;

    public RecentViewViewModel() {
    }

    protected RecentViewViewModel(Parcel in) {
        recentView = in.readParcelable(RecentView.class.getClassLoader());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(recentView, flags);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<RecentViewViewModel> CREATOR = new Creator<RecentViewViewModel>() {
        @Override
        public RecentViewViewModel createFromParcel(Parcel in) {
            return new RecentViewViewModel(in);
        }

        @Override
        public RecentViewViewModel[] newArray(int size) {
            return new RecentViewViewModel[size];
        }
    };

    public RecentView getRecentView() {
        return recentView;
    }

    public void setRecentView(RecentView recentView) {
        this.recentView = recentView;
    }
}

File: features/transaction/checkout/src/main/java/com/tokopedia/checkout/view/feature/emptycart2/viewmodel/RecentViewViewModel.kt
package com.tokopedia.checkout.view.feature.emptycart2.viewmodel

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.tokopedia.checkout.domain.datamodel.recentview.GqlRecentViewResponse
import com.tokopedia.checkout.domain.datamodel.recentview.RecentView
import com.tokopedia.checkout.domain.usecase.GetRecentViewUseCase
import com.tokopedia.graphql.data.model.GraphqlResponse
import com.tokopedia.usecase.coroutines.Fail
import com.tokopedia.usecase.coroutines.Result
import com.tokopedia.usecase.coroutines.Success
import com.tokopedia.user.session.UserSessionInterface
import rx.Subscriber
import javax.inject.Inject

/**
 * Created by Irfan Khoirul on 2019-05-20.
 */

class RecentViewViewModel @Inject constructor(private val getRecentViewUseCase: GetRecentViewUseCase,
                                              private val userSessionInterface: UserSessionInterface) : ViewModel() {

    val recentViewData = MutableLiveData<Result<MutableList<RecentView>>>()

    fun unsubscribeSubscription() {
        getRecentViewUseCase.unsubscribe()
    }

    fun getRecentView() {
        getRecentViewUseCase.createObservable(userSessionInterface.userId.toInt(), object : Subscriber<GraphqlResponse>() {
            override fun onCompleted() {

            }

            override fun onError(e: Throwable) {
                recentViewData.value = Fail(RuntimeException())
            }

            override fun onNext(graphqlResponse: GraphqlResponse) {
                if (graphqlResponse.getData<Any>(GqlRecentViewResponse::class.java) != null) {
                    val gqlRecentViewResponse = graphqlResponse.getData<GqlRecentViewResponse>(GqlRecentViewResponse::class.java)
                    if (gqlRecentViewResponse != null && gqlRecentViewResponse.gqlRecentView != null &&
                            gqlRecentViewResponse.gqlRecentView.recentViewList != null &&
                            gqlRecentViewResponse.gqlRecentView.recentViewList.size > 0) {
                        if (gqlRecentViewResponse.gqlRecentView.recentViewList.size > 2) {
                            recentViewData.value = Success(gqlRecentViewResponse.gqlRecentView.recentViewList.subList(0, 2))
                        } else {
                            recentViewData.value = Success(gqlRecentViewResponse.gqlRecentView.recentViewList)
                        }
                    } else {
                        recentViewData.value = Fail(RuntimeException())
                    }
                } else {
                    recentViewData.value = Fail(RuntimeException())
                }
            }
        })
    }

}